File: benchmarks/fannkuch/fannkuch.js
function fannkuch(n) {
    const perm = new Array(n);
    const perm1 = new Array(n);
    const count = new Array(n);

    for (let i = 0; i < n; i++) {
        perm1[i] = i;
    }

    let maxFlips = 0;
    let checksum = 0;
    let sign = 1;
    let r = n;

    while (true) {
        // Generate next permutation
        while (r !== 1) {
            count[r - 1] = r;
            r--;
        }

        // Copy perm1 to perm
        for (let i = 0; i < n; i++) {
            perm[i] = perm1[i];
        }

        // Count flips
        let flips = 0;
        while (perm[0] !== 0) {
            const k = perm[0] + 1;
            // Reverse first k elements
            for (let i = 0, j = k - 1; i < j; i++, j--) {
                const tmp = perm[i];
                perm[i] = perm[j];
                perm[j] = tmp;
            }
            flips++;
        }

        if (flips > maxFlips) {
            maxFlips = flips;
        }
        checksum += sign * flips;
        sign = -sign;

        // Next permutation
        while (true) {
            if (r === n) {
                console.log(checksum);
                return maxFlips;
            }
            // Rotate perm1[0..r]
            const p0 = perm1[0];
            for (let i = 0; i < r; i++) {
                perm1[i] = perm1[i + 1];
            }
            perm1[r] = p0;

            count[r]--;
            if (count[r] > 0) {
                break;
            }
            r++;
        }
    }
}

const n = 9;
const result = fannkuch(n);
console.log(`Pfannkuchen(${n}) = ${result}`);

File: benchmarks/fannkuch/Fannkuch.java
public class Fannkuch {
    static int fannkuch(int n) {
        int[] perm = new int[n];
        int[] perm1 = new int[n];
        int[] count = new int[n];

        for (int i = 0; i < n; i++) {
            perm1[i] = i;
        }

        int maxFlips = 0;
        int checksum = 0;
        int sign = 1;
        int r = n;

        while (true) {
            // Generate next permutation
            while (r != 1) {
                count[r - 1] = r;
                r--;
            }

            // Copy perm1 to perm
            System.arraycopy(perm1, 0, perm, 0, n);

            // Count flips
            int flips = 0;
            while (perm[0] != 0) {
                int k = perm[0] + 1;
                // Reverse first k elements
                for (int i = 0, j = k - 1; i < j; i++, j--) {
                    int tmp = perm[i];
                    perm[i] = perm[j];
                    perm[j] = tmp;
                }
                flips++;
            }

            if (flips > maxFlips) {
                maxFlips = flips;
            }
            checksum += sign * flips;
            sign = -sign;

            // Next permutation
            while (true) {
                if (r == n) {
                    System.out.println(checksum);
                    return maxFlips;
                }
                // Rotate perm1[0..r]
                int p0 = perm1[0];
                for (int i = 0; i < r; i++) {
                    perm1[i] = perm1[i + 1];
                }
                perm1[r] = p0;

                count[r]--;
                if (count[r] > 0) {
                    break;
                }
                r++;
            }
        }
    }

    public static void main(String[] args) {
        int n = 9;
        int result = fannkuch(n);
        System.out.printf("Pfannkuchen(%d) = %d%n", n, result);
    }
}
